File: Knapsack.java
import java.util.Scanner;

public class Knapsack {

    public static int knapsack(int capacity, int[] weights, int[] values, int n) {
        int[][] dp = new int[n + 1][capacity + 1];

        for (int i = 0; i <= n; i++) {
            for (int w = 0; w <= capacity; w++) {
                if (i == 0 || w == 0) {
                    dp[i][w] = 0;
                } else if (weights[i - 1] <= w) {
                    dp[i][w] = Math.max(values[i - 1] + dp[i - 1][w - weights[i - 1]], dp[i - 1][w]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }
        return dp[n][capacity];
    }

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter number of items: ");
            int n = scanner.nextInt();

            int[] values = new int[n];
            int[] weights = new int[n];

            for (int i = 0; i < n; i++) {
                System.out.print("Enter value for item " + (i + 1) + ": ");
                values[i] = scanner.nextInt();
                System.out.print("Enter weight for item " + (i + 1) + ": ");
                weights[i] = scanner.nextInt();
            }

            System.out.print("Enter knapsack capacity: ");
            int capacity = scanner.nextInt();

            int maxValue = knapsack(capacity, weights, values, n);
            System.out.println("Maximum value achievable: " + maxValue);
        }
    }
}

File: knapsack.py
import csv

def knapsack(capacity, weights, values):
    n = len(values)
    # Create a DP table with (n+1) rows and (capacity+1) columns
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]
    
    for i in range(n + 1):
        for w in range(capacity + 1):
            if i == 0 or w == 0:
                dp[i][w] = 0
            elif weights[i-1] <= w:
                dp[i][w] = max(values[i-1] + dp[i-1][w - weights[i-1]], dp[i-1][w])
            else:
                dp[i][w] = dp[i-1][w]
    return dp[n][capacity]

def read_knapsack_items(filename):
    values = []
    weights = []
    
    with open(filename, 'r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            values.append(int(row['value']))
            weights.append(int(row['weight']))
    return values, weights

if __name__ == "__main__":
    # Read items from CSV
    values, weights = read_knapsack_items('knapsack.csv')
    
    # Get knapsack capacity from user
    capacity = int(input("Enter knapsack capacity: "))
    
    # Calculate maximum value
    max_value = knapsack(capacity, weights, values)
    
    # Display result
    print(f"\nItems loaded from CSV: {len(values)}")
    print(f"Knapsack capacity: {capacity}")
    print(f"Maximum achievable value: {max_value}")